Project flags, for global config defaults + behavior change migrations

jtcohen6 opened the pull request:
resolves #4366
resolves #4709
resolves #4787
resolves #2894

## What are you changing in this pull request and why?

- Reframe "global configs" &rarr; "flags"
- Settable in `dbt_project.yml` instead of `profiles.yml`
- Subcategory of flags for legacy behaviors / behavior change migrations

<details>
<summary>

### Previews
</summary>

Most important changes:
- [website/docs/reference/global-configs/about-global-configs.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/about-global-configs)
- [website/docs/reference/global-configs/project-flags.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/project-flags)
- [website/docs/reference/dbt_project.yml.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/dbt_project.yml)
- [website/docs/reference/global-configs/legacy-behaviors.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/legacy-behaviors)
- [website/docs/docs/core/connect-data-platform/profiles.yml.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/docs/core/connect-data-platform/profiles.yml)

Other new pages:
- [website/docs/reference/global-configs/record-timing-info.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/record-timing-info)
- [website/docs/reference/global-configs/indirect-selection.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/indirect-selection)

Other significant edits:
- [website/docs/reference/global-configs/logs.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/logs)
- [website/docs/reference/global-configs/parsing.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/parsing)
- [website/docs/reference/global-configs/version-compatibility.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/version-compatibility)
- [website/docs/reference/global-configs/json-artifacts.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/json-artifacts)
- website/sidebars.js
- website/vercel.json
- website/dbt-versions.js

Edited links:
- [website/docs/docs/dbt-versions/core-upgrade/08-upgrading-to-v1.2.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/docs/dbt-versions/core-upgrade/upgrading-to-v1.2)
- [website/docs/docs/dbt-versions/core-upgrade/10-upgrading-to-v1.0.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/docs/dbt-versions/core-upgrade/upgrading-to-v1.0)
- [website/docs/docs/dbt-cloud-apis/sl-manifest.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/docs/dbt-cloud-apis/sl-manifest)
- [website/docs/reference/artifacts/dbt-artifacts.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/artifacts/dbt-artifacts)
- [website/docs/reference/events-logging.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/events-logging)
- [website/docs/reference/parsing.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/parsing)
- [website/docs/reference/project-configs/clean-targets.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/project-configs/clean-targets)
- [website/docs/reference/resource-configs/severity.md](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/resource-configs/severity)

Deleted:
- website/docs/reference/global-cli-flags.md
- website/docs/reference/global-configs/yaml-configurations.md
- website/docs/reference/project-configs/log-path.md
- website/docs/reference/project-configs/target-path.md

Minor edits (moved and re-titled):
- website/docs/reference/global-configs/command-line-options.md

Other:
- .github/pull_request_template.md

</details>

## Checklist

Adding or removing pages (delete if not applicable):
- [x] Add/remove page in `website/sidebars.js`
- [x] Provide a unique filename for new pages
- [x] Add an entry for deleted pages in `website/vercel.json`
- [x] Run link testing locally with `npm run build` to update the links that point to deleted pages

jtcohen6 committed:
First draft: global configs --> flags

dbeatty10 commented on website/docs/reference/global-configs/legacy-behaviors.md:
```diff
@@ -0,0 +1,36 @@
+# dbt_project.yml
+
+on-run-start:
+  - '{{ ... if flags.WHICH != 'freshness' }}'
```
It would be awesome if we do https://github.com/dbt-labs/docs.getdbt.com/issues/2825 !

jtcohen6 replied:
I really thought about it... but I think I need to leave it for another PR

dbeatty10 commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,57 @@
+- **[Environment variables](https://docs.getdbt.com/reference/global-configs/environment-variable-configs):** Define behavior that should be different in different runtime environments (development vs. production vs. [continuous integration](https://docs.getdbt.com/docs/deploy/continuous-integration), or different for different users in development (based on personal preferences).
+- **[CLI options](https://docs.getdbt.com/reference/global-configs/command-line-options):** Define behavior specific to _this invocation_. Supported for all dbt commands.
+
+The precedence order is CLI > Env Var > Project.
```
In this particular case, it's not obvious if the `>` are "greater (precedence) than" or if they are more like directional arrows.

Generally, it would be nice to have consistent and clear ways to communicate the precedence of various things. See https://github.com/dbt-labs/docs.getdbt.com/issues/4371 for a request for some standardization across our docs.

dbeatty10 commented on website/docs/reference/dbt_project.yml.md:
```diff
@@ -52,6 +52,9 @@ The following example is a list of all available configurations in the `dbt_proj
 
+[flags](/reference/global-configs/project-flags):
+  <global-configs>
+
```
It would be nice if `<global-configs>` linked to a place with examples similar to each of these:
- [`<model-configs>`](https://docs.getdbt.com/reference/model-configs#general-configurations)
- [`seed-configs`](https://docs.getdbt.com/reference/seed-configs#seed-specific-configurations)
- [`snapshot-configs`](https://docs.getdbt.com/reference/snapshot-configs#snapshot-specific-configurations)
- [`source-configs`](https://docs.getdbt.com/reference/source-configs#general-configurations)
- [`test-configs`](https://docs.getdbt.com/reference/data-test-configs#data-test-specific-configurations)

Maybe a specific section within [global-configs.md](https://docs.getdbt.com/reference/global-configs/about-global-configs) that showed all these generic structure of setting flags in `dbt_project.yml`?

That could help fill in some of the gaps related to https://github.com/dbt-labs/docs.getdbt.com/issues/4366.

dbeatty10 commented on website/docs/reference/global-configs/project-flags.md:
````diff
@@ -1,9 +1,26 @@
+```yaml
+
+flags:
+  <global_config>: <value>
````
See earlier comment relating to linking `<global-configs>` to examples that enumerate all the names and options.

dbeatty10 commented on website/docs/reference/global-configs/project-flags.md:
```diff
@@ -1,9 +1,26 @@
+sidebar: "Project flags"
 ---
 
+<File name='dbt_project.yml.yml'>
```
Little Caesars ad folks must've gotta hold on this file name -- 'cause you can never have enough YAML. 

```suggestion
<File name='dbt_project.yml'>
```

dbeatty10 reviewed (commented):
You are a tour de force 🤩

Didn't look at any previews yet, but read through the code and left some comments below.

Something we should consider handling either within this PR or a separate one: https://github.com/dbt-labs/docs.getdbt.com/issues/4366

See https://github.com/dbt-labs/docs.getdbt.com/issues/4366#issuecomment-1889503363 for one idea to tie together CLI flags and environment variable names. Unfortunately, it wouldn't also tie together flag settings within `dbt_project.yml` too.

jtcohen6 committed:
Fix broken links

jtcohen6 committed:
PR feedback

jtcohen6 reviewed (commented):
> Something we should consider handling either within this PR or a separate one: https://github.com/dbt-labs/docs.getdbt.com/issues/4366

I gave it a go! It's not the prettiest, but I stand by the information in there. Let me know what you think.

https://docs-getdbt-7fp98qtm3-dbt-labs.vercel.app/reference/global-configs/about-global-configs#available-flags

runleonarun committed:
addinb new indirect selection pages

runleonarun committed:
Merge branch 'current' into indirect-selection4773

runleonarun committed:
adding to sidebar

runleonarun committed:
closing file

runleonarun committed:
Merge branch 'indirect-selection4773' of github.com:dbt-labs/docs.getdbt.com into indirect-selection4773

runleonarun committed:
fixing the env var example

runleonarun committed:
Merge branch 'current' into indirect-selection4773

runleonarun committed:
fix example sentences

runleonarun committed:
Update website/dbt-versions.js

runleonarun commented on .github/pull_request_template.md:
```diff
@@ -19,5 +19,5 @@ Uncomment when publishing docs for a prerelease version of dbt:
 - [ ] Add/remove page in `website/sidebars.js`
 - [ ] Provide a unique filename for new pages
-- [ ] Add an entry for deleted pages in `website/static/_redirects`
+- [ ] Add an entry for deleted pages in `website/vercel.json`
```
Nice catch 🙇🏻

runleonarun committed:
Apply suggestions from code review

Co-authored-by: Ly Nguyen <107218380+nghi-ly@users.noreply.github.com>

runleonarun committed:
Apply suggestions from code review

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
+In dbt, "flags" (also called "global configs") are configurations for fine-tuning _how_ dbt runs your project. They differ from [resource-specific configs](/reference/configs-and-properties) that tell dbt about _what_ to run.
 
-Global configs control things like the visual output of logs, the manner in which dbt parses your project, and what to do when dbt finds a version mismatch or a failing model. These configs are "global" because they are available for all dbt commands, and because they can be set for all projects running on the same machine or in the same environment.
+Flags control things like the visual output of logs, whether to treat specific warning messages as errors, or whether to "fail fast" after encountering the first error. These flags are "global" because they are available for all dbt commands, and because they can be set in multiple places.
```
```suggestion
Flags control things like the visual output of logs, whether to treat specific warning messages as errors, or whether to "fail fast" after encountering the first error. Flags are "global" configs because they are available for all dbt commands and they can be set in multiple places.
```

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
+Flags control things like the visual output of logs, whether to treat specific warning messages as errors, or whether to "fail fast" after encountering the first error. These flags are "global" because they are available for all dbt commands, and because they can be set in multiple places.
 
-### Global config precedence
+There is significant overlap between dbt's flags and dbt's command line options, but they are not the same:
```
```suggestion
There is a significant overlap between dbt's flags and dbt's command line options, but there are differences:
```

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
 
-### Global config precedence
+There is significant overlap between dbt's flags and dbt's command line options, but they are not the same:
+- Certain flags can only be set in `dbt_project.yml`, and cannot be overridden for specific invocations via CLI option.
```
```suggestion
- Certain flags can only be set in `dbt_project.yml` and cannot be overridden for specific invocations via CLI options.
```

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
+| [target](/docs/core/connect-data-platform/connection-profiles#about-profiles) | path | None (uses `target/`) | ❌ | `DBT_TARGET` | `--target` | ❌ |
+| [use_colors_file](/reference/global-configs/logs#color) | boolean | True | ✅ | `DBT_USE_COLORS_FILE` | `--use-colors-file`, `--no-use-colors-file` | ❌ |
+| [use_colors](/reference/global-configs/print-output#print-color) | boolean | True | ✅ | `DBT_USE_COLORS` | `--use-colors`, `--no-use-colors` | ❌ |
+| [use_experimental_parser](/reference/global-configs/parsing#experimental-parser) | boolean | False | ✅ | `DBT_USE_EXPERIMENTAL_PARSER` | `--use-experimental-parser`, `--no-use-experimental-parser` | ❌ |
```
This leads to a dead-end, and I see that we're not version-blocking the deleted text. Should we just remove the entry and the result?

dbeatty10 replied:
Good call.

Split this feedback into a separate issue to tackle as a follow-up: https://github.com/dbt-labs/docs.getdbt.com/issues/4990

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
 
-Each config is prioritized over the previous one. For example, if all three are provided, then the CLI flag takes precedence.
+There are multiple ways of setting flags, which depend on the use case:
+- **[Project-level `flags` (`dbt_project.yml`)](https://docs.getdbt.com/reference/global-configs/project-flags):** Define version-controlled defaults for everyone running this project. Preserve legacy behaviors until their slated deprecation.
```
Did you intend to use the full URL for these links? That will open them in a new tab rather than in the current tab as the other links on this page are configured.

dbeatty10 replied:
It appears like the full URLs were an accident that were inherited from a previous PR.

So I'm going to remove the hard-coded link in favor of relative ones.

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
-Each config is prioritized over the previous one. For example, if all three are provided, then the CLI flag takes precedence.
+There are multiple ways of setting flags, which depend on the use case:
+- **[Project-level `flags` (`dbt_project.yml`)](https://docs.getdbt.com/reference/global-configs/project-flags):** Define version-controlled defaults for everyone running this project. Preserve legacy behaviors until their slated deprecation.
+- **[Environment variables](https://docs.getdbt.com/reference/global-configs/environment-variable-configs):** Define behavior that should be different in different runtime environments (development vs. production vs. [continuous integration](https://docs.getdbt.com/docs/deploy/continuous-integration), or different for different users in development (based on personal preferences).
```
```suggestion
- **[Environment variables](https://docs.getdbt.com/reference/global-configs/environment-variable-configs):** Define different behavior in different runtime environments (development vs. production vs. [continuous integration](https://docs.getdbt.com/docs/deploy/continuous-integration), or different behavior for different users in development (based on personal preferences).
```

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
+- **[Environment variables](https://docs.getdbt.com/reference/global-configs/environment-variable-configs):** Define behavior that should be different in different runtime environments (development vs. production vs. [continuous integration](https://docs.getdbt.com/docs/deploy/continuous-integration), or different for different users in development (based on personal preferences).
+- **[CLI options](https://docs.getdbt.com/reference/global-configs/command-line-options):** Define behavior specific to _this invocation_. Supported for all dbt commands.
+
+The most specific setting "wins." If you set the same flag in all three ways, the CLI option will take precedence, followed by the environment variable, followed lastly by the value in `dbt_project.yml`. If you set the flag in none of those places, it will use the default value defined within dbt.
```
```suggestion
The most specific setting "wins." If you set the same flag in all three places, the CLI option will take precedence, followed by the environment variable, and finally, the value in `dbt_project.yml`. If you set the flag in none of those places, it will use the default value defined within dbt.
```

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
````diff
@@ -1,18 +1,97 @@
+```
+
+There are two categories of exceptions:
+1. **Flags setting file paths:** Flags for file paths that are relevant to runtime execution (e.g. `--log-path` or `--state`) cannot be set in `dbt_project.yml`. To override defaults, pass CLI options or set environment variables (`DBT_LOG_PATH`, `DBT_STATE`). Flags that tell dbt where to find project resources (e.g. `model-paths`) are set in `dbt_project.yml`, but as a top-level key, outside the ...
````
```suggestion
1. **Flags setting file paths:** Flags for file paths that are relevant to runtime execution (for example, `--log-path` or `--state`) cannot be set in `dbt_project.yml`. To override defaults, pass CLI options or set environment variables (`DBT_LOG_PATH`, `DBT_STATE`). Flags that tell dbt where to find project resources (for example, `model-paths`) are set in `dbt_project.yml`, but as a top-level key, outside the `flags` dictionary; these configs are expected to be fully static and never vary based on the command or execution environment.
```

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
+
+There are two categories of exceptions:
+1. **Flags setting file paths:** Flags for file paths that are relevant to runtime execution (e.g. `--log-path` or `--state`) cannot be set in `dbt_project.yml`. To override defaults, pass CLI options or set environment variables (`DBT_LOG_PATH`, `DBT_STATE`). Flags that tell dbt where to find project resources (e.g. `model-paths`) are set in `dbt_project.yml`, but as a top-level key, outside the ...
+2. Flags opting into legacy dbt behaviors can _only_ be defined in `dbt_project.yml`. These are intended to be set in version control, and migrated via pull/merge request. Their values should not diverge indefinitely across invocations, environments, or users.
```
```suggestion
2. Flags opting into legacy dbt behaviors can _only_ be defined in `dbt_project.yml`. These are intended to be set in version control and migrated via pull/merge request. Their values should not diverge indefinitely across invocations, environments, or users.
```

matthewshaver commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
+
+Because the values of `flags` can differ across invocations, we strongly advise against using `flags` as an input to configurations or dependencies (`ref` + `source`) that dbt resolves [during parsing](https://docs.getdbt.com/reference/parsing#known-limitations).
+
+### Available flags
```
```suggestion
## Available flags
```

matthewshaver commented on website/docs/reference/global-configs/command-line-options.md:
```diff
@@ -1,7 +1,7 @@
-sidebar: "Command line flags"
+title: "Command line options"
+id: "command-line-options"
+sidebar: "Command line options"
```
Should we change the use of `flags` in the body of this doc to `options`?

dbeatty10 replied:
["Flags"](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/about-global-configs) is the unifying description we're giving, so the body of this doc can stay as-is.

### [More detail](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/about-global-configs#setting-flags)

There are multiple ways of setting flags, which depend on the use case:

- **[Project-level `flags` (`dbt_project.yml`)](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/project-flags):** Define version-controlled defaults for everyone running this project. Preserve legacy behaviors until their slated deprecation.
- **[Environment variables](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/environment-variable-configs):** Define different behavior in different runtime environments (development vs. production vs. [continuous integration](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/docs/deploy/continuous-integration), or different behavior for different users in development (based on personal preferences).
- **[CLI options](https://docs-getdbt-com-git-jerco-flags-behavior-changes-dbt-labs.vercel.app/reference/global-configs/command-line-options):** Define behavior specific to _this invocation_. Supported for all dbt commands.

matthewshaver commented on website/docs/reference/global-configs/json-artifacts.md:
```diff
@@ -4,7 +4,9 @@ id: "json-artifacts"
-The `WRITE_JSON` config determines whether dbt writes JSON artifacts (eg. `manifest.json`, `run_results.json`) to the `target/` directory. JSON serialization can be slow, and turning this flag off _might_ make invocations of dbt faster. Alternatively, you might disable this config if you want to perform a dbt operation and avoid overwriting artifacts from a previous run step.
+### Write JSON artifacts
+
+The `WRITE_JSON` config determines whether dbt writes [JSON artifacts](/reference/artifacts/dbt-artifacts) (eg. `manifest.json`, `run_results.json`) to the `target/` directory. JSON serialization can be slow, and turning this flag off _might_ make invocations of dbt faster. Alternatively, you might disable this config if you want to perform a dbt operation and avoid overwriting artifacts from a p ...
```
```suggestion
The `WRITE_JSON` config determines whether dbt writes [JSON artifacts](/reference/artifacts/dbt-artifacts) (for example, `manifest.json`, `run_results.json`) to the `target/` directory. JSON serialization can be slow, and turning this flag off _might_ make invocations of dbt faster. Alternatively, you can disable this config to perform a dbt operation and avoid overwriting artifacts from a previous run step.
```

matthewshaver commented on website/docs/reference/global-configs/legacy-behaviors.md:
```diff
@@ -0,0 +1,36 @@
+Most flags exist to configure runtime behaviors with multiple valid choices. The right choice may vary based on the environment, user preference, or the specific invocation.
+
+Another category of flags provides existing projects with a migration window for runtime behaviors that are changing in newer releases of dbt. These flags help us achieve a balance between these goals, which can otherwise be in tension:
+1. Providing a better, more sensible, and more consistent default behaviour for new users/projects going forward
```
```suggestion
- Providing a better, more sensible, and more consistent default behavior for new users/projects.
```

matthewshaver commented on website/docs/reference/global-configs/legacy-behaviors.md:
```diff
@@ -0,0 +1,36 @@
+
+Most flags exist to configure runtime behaviors with multiple valid choices. The right choice may vary based on the environment, user preference, or the specific invocation.
+
+Another category of flags provides existing projects with a migration window for runtime behaviors that are changing in newer releases of dbt. These flags help us achieve a balance between these goals, which can otherwise be in tension:
```
```suggestion
Another category of flags provides existing projects with a migration window for runtime behaviors that are changing in newer releases of dbt. These flags help us achieve a balance between these goals, which can otherwise be in tension, by:
```

matthewshaver commented on website/docs/reference/global-configs/legacy-behaviors.md:
```diff
@@ -0,0 +1,36 @@
+
+Another category of flags provides existing projects with a migration window for runtime behaviors that are changing in newer releases of dbt. These flags help us achieve a balance between these goals, which can otherwise be in tension:
+1. Providing a better, more sensible, and more consistent default behaviour for new users/projects going forward
+2. Providing a migration window for existing users/projects — nothing changes overnight without warning
```
```suggestion
- Providing a migration window for existing users/projects &mdash; nothing changes overnight without warning.
```

matthewshaver commented on website/docs/reference/global-configs/legacy-behaviors.md:
```diff
@@ -0,0 +1,36 @@
+Another category of flags provides existing projects with a migration window for runtime behaviors that are changing in newer releases of dbt. These flags help us achieve a balance between these goals, which can otherwise be in tension:
+1. Providing a better, more sensible, and more consistent default behaviour for new users/projects going forward
+2. Providing a migration window for existing users/projects — nothing changes overnight without warning
+3. Maintainability of dbt software. Every fork in behaviour requires additional testing & cognitive overhead that slows future development. These flags exist to facilitate migration from "current" to "better," not to stick around forever.
```
```suggestion
- Providing maintainability of dbt software. Every fork in behavior requires additional testing & cognitive overhead that slows future development. These flags exist to facilitate migration from "current" to "better," not to stick around forever.
```

matthewshaver commented on website/docs/reference/global-configs/legacy-behaviors.md:
```diff
@@ -0,0 +1,36 @@
+These flags can _only_ be set in `dbt_project.yml`. They configure behaviors closely tied to project code. They should be defined in version control, and changed via pull/merge request, with the same testing and peer review.
+
+## Behaviors
+
```
Could we get a little descriptor in here? We try to avoid back-to-back headings.

matthewshaver replied:
Will address is a follow up!

jtcohen6 replied:
@matthewshaver Very happy to have you push commits directly to this branch (which I see you're doing). I can give another read-through with the changes once they're final.

matthewshaver commented on website/docs/reference/global-configs/project-flags.md:
```diff
@@ -0,0 +1,47 @@
+
+</File>
+
+See the [table of all flags](/reference/global-configs/about-global-configs#available-flags) to see which global configs are available for setting in `dbt_project.yml`.
```
```suggestion
Reference the [table of all flags](/reference/global-configs/about-global-configs#available-flags) to see which global configs are available for setting in `dbt_project.yml`.
```

matthewshaver commented on website/docs/reference/global-configs/project-flags.md:
```diff
@@ -0,0 +1,47 @@
+<VersionBlock lastVersion="1.7">
+
+:::warning Deprecated functionality
+In older versions of dbt, custom default values of flags (global configs) were set in `profiles.yml`. Starting in v1.8, these configs should be set in `dbt_project.yml` instead.
```
```suggestion
In older versions of dbt, custom default values of flags (global configs) were set in `profiles.yml`. Starting in v1.8, these configs are set in `dbt_project.yml` instead.
```

matthewshaver commented on website/docs/reference/global-configs/project-flags.md:
```diff
@@ -0,0 +1,47 @@
+In older versions of dbt, custom default values of flags (global configs) were set in `profiles.yml`. Starting in v1.8, these configs should be set in `dbt_project.yml` instead.
+:::
+
+For most global configurations, you can set "user profile" configurations in the `config:` block of `profiles.yml`. This style of configuration sets default values for all projects using this profile directory—usually, all projects running on your local machine.
```
```suggestion
For most global configurations, you can set "user profile" configurations in the `config:` block of `profiles.yml`. This style of configuration sets default values for all projects using this profile directory &mdash; usually, all projects running on your local machine.
```

matthewshaver commented on website/docs/reference/global-configs/record-timing-info.md:
```diff
@@ -0,0 +1,17 @@
+id: "record-timing-info"
+---
+
+The `-r` or `--record-timing-info` flag saves performance profiling information to a file. This file can be visualized with `snakeviz` to understand the performance characteristics of a dbt invocation
```
```suggestion
The `-r` or `--record-timing-info` flag saves performance profiling information to a file. This file can be visualized with `snakeviz` to understand the performance characteristics of a dbt invocation.
```

matthewshaver commented on website/docs/reference/global-configs/version-compatibility.md:
```diff
@@ -4,7 +4,7 @@ id: "version-compatibility"
 ---
 
-Projects are recommended to set [dbt version requirements](/reference/project-configs/require-dbt-version), especially if they use features that are newer, or which may break in future versions of dbt Core. By default, if you run a project with an incompatible dbt version, dbt will raise an error.
+In the first several years of dbt Core's development, breaking changes were more common. For this reason, we encouraged to set [dbt version requirements](/reference/project-configs/require-dbt-version) — especially if they use features that are newer, or which may break in future versions of dbt Core. By default, if you run a project with an incompatible dbt version, dbt will raise an error.
```
```suggestion
In the first several years of dbt Core's development, breaking changes were more common. For this reason, we encouraged to set [dbt version requirements](/reference/project-configs/require-dbt-version) — especially if they use features that are newer, or which may break in future versions of dbt Core. By default, if you run a project with an incompatible dbt version, dbt will raise an error.
```
```suggestion
For the first several years of dbt Core's development, breaking changes were more common. For this reason, we encouraged setting [dbt version requirements](/reference/project-configs/require-dbt-version) &mdash; especially if they use features that are newer or which may break in future versions of dbt Core. By default, if you run a project with an incompatible dbt version, dbt will raise an error.
```

matthewshaver commented on website/docs/reference/global-configs/record-timing-info.md:
```diff
@@ -0,0 +1,17 @@
+---
+title: "Record timing info"
```
This one doesn't appear to be on the sidebar. Is that by design?

dbeatty10 replied:
Wow, eagle eyes! 🦅 

Added to the sidebar in https://github.com/dbt-labs/docs.getdbt.com/pull/4814/commits/b29c9d58c6c3129d011cd2659471a80c5c1c168c

runleonarun committed:
Update website/docs/reference/global-configs/indirect-selection.md

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

dbeatty10 commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
 
-Each config is prioritized over the previous one. For example, if all three are provided, then the CLI flag takes precedence.
+There are multiple ways of setting flags, which depend on the use case:
+- **[Project-level `flags` (`dbt_project.yml`)](https://docs.getdbt.com/reference/global-configs/project-flags):** Define version-controlled defaults for everyone running this project. Preserve legacy behaviors until their slated deprecation.
```
```suggestion
- **[Project-level `flags` (`dbt_project.yml`)](/reference/global-configs/project-flags):** Define version-controlled defaults for everyone running this project. Preserve legacy behaviors until their slated deprecation.
```

dbeatty10 committed:
Apply suggestions from code review

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

dbeatty10 commented on website/docs/reference/global-configs/legacy-behaviors.md:
```diff
@@ -0,0 +1,36 @@
+## Behaviors
+
+### `source_freshness_run_project_hooks`
+- Should the 'dbt source freshness' command include [on-run-start / on-run-end hooks](https://docs.getdbt.com/reference/project-configs/on-run-start-on-run-end)?
```
```suggestion
- Should the 'dbt source freshness' command include [on-run-start / on-run-end hooks](/reference/project-configs/on-run-start-on-run-end)?
```

dbeatty10 commented on website/docs/reference/global-configs/about-global-configs.md:
````diff
@@ -1,18 +1,97 @@
+  - '{{ log("I will stop at the first sign of trouble", info = true) if flags.FAIL_FAST }}'
+```
+
+Because the values of `flags` can differ across invocations, we strongly advise against using `flags` as an input to configurations or dependencies (`ref` + `source`) that dbt resolves [during parsing](https://docs.getdbt.com/reference/parsing#known-limitations).
````
```suggestion
Because the values of `flags` can differ across invocations, we strongly advise against using `flags` as an input to configurations or dependencies (`ref` + `source`) that dbt resolves [during parsing](/reference/parsing#known-limitations).
```

dbeatty10 commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
+
+### Accessing flags
+
+Custom user-defined logic, written in Jinja, can check the values of flags using [the `flags` context variable](https://docs.getdbt.com/reference/dbt-jinja-functions/flags).
```
```suggestion
Custom user-defined logic, written in Jinja, can check the values of flags using [the `flags` context variable](/reference/dbt-jinja-functions/flags).
```

dbeatty10 commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,97 @@
+There are multiple ways of setting flags, which depend on the use case:
+- **[Project-level `flags` (`dbt_project.yml`)](https://docs.getdbt.com/reference/global-configs/project-flags):** Define version-controlled defaults for everyone running this project. Preserve legacy behaviors until their slated deprecation.
+- **[Environment variables](https://docs.getdbt.com/reference/global-configs/environment-variable-configs):** Define different behavior in different runtime environments (development vs. production vs. [continuous integration](https://docs.getdbt.com/docs/deploy/continuous-integration), or different behavior for different users in development (based on personal preferences).
+- **[CLI options](https://docs.getdbt.com/reference/global-configs/command-line-options):** Define behavior specific to _this invocation_. Supported for all dbt commands.
```
```suggestion
- **[Environment variables](/reference/global-configs/environment-variable-configs):** Define different behavior in different runtime environments (development vs. production vs. [continuous integration](/docs/deploy/continuous-integration), or different behavior for different users in development (based on personal preferences).
- **[CLI options](/reference/global-configs/command-line-options):** Define behavior specific to _this invocation_. Supported for all dbt commands.
```

dbeatty10 committed:
Use relative links instead of absolute

matthewshaver committed:
Apply suggestions from code review

Editorial changes

matthewshaver committed:
Update website/docs/reference/global-configs/about-global-configs.md

runleonarun committed:
Merge branch 'jerco/flags-behavior-changes' into indirect-selection4773

runleonarun committed:
Merge branch 'current' into jerco/flags-behavior-changes

runleonarun committed:
Merge branch 'jerco/flags-behavior-changes' into indirect-selection4773

runleonarun committed:
Update profiles.yml.md

runleonarun committed:
Apply suggestions from code review

Co-authored-by: Matt Shaver <60105315+matthewshaver@users.noreply.github.com>

runleonarun committed:
adding backindirect selection

runleonarun commented:
@jtcohen6 I'm trying to translate the changes in this PR over to my PR on indirect_selection, and am looking at the [Configs properties, what are they?](https://docs-getdbt-50nqy6oil-dbt-labs.vercel.app/reference/configs-and-properties) page. Should we be making any updates to that page to align with the global configs are now being called flags change?

jtcohen6 commented:
> 1. Should we be making any updates to [that page](https://docs-getdbt-50nqy6oil-dbt-labs.vercel.app/reference/configs-and-properties) to align with the global configs are now being called flags change?

No - those are resource-level configs, which modify the definition and behavior of specific resources in the project (~nodes in the DAG), e.g. "This model is a table, and it has tag `pii`.")

We should keep calling them that, and avoid as much confusion as possible with `flags` (aka / fka "global" configs), which affect _how_ dbt runs, e.g. "In CI environments, stop on the first error."

2. Should we move the global configs [captured here in the profiles.yml](https://docs-getdbt-50nqy6oil-dbt-labs.vercel.app/docs/core/connect-data-platform/profiles.yml) over to the project file?

Yes! Starting in v1.8, we are deprecating support for these being defined in `profiles.yml`.

dbeatty10 commented on website/sidebars.js:
```diff
@@ -932,27 +930,39 @@ const sidebarSettings = {
+                "reference/global-configs/json-artifacts",
+                "reference/global-configs/legacy-behaviors",
+                "reference/global-configs/parsing",
+                "reference/global-configs/print-output",
+                "reference/global-configs/usage-stats",
```
Adding per https://github.com/dbt-labs/docs.getdbt.com/pull/4814/files#r1500210472

```suggestion
                "reference/global-configs/print-output",
                "reference/global-configs/record-timing-info",
                "reference/global-configs/usage-stats",
```

dbeatty10 committed:
Update website/sidebars.js to include record-timing-info

runleonarun committed:
Apply suggestions from code review

Co-authored-by: Doug Beatty <44704949+dbeatty10@users.noreply.github.com>

matthewshaver committed:
Updating profiles.yml

matthewshaver committed:
Merge branch 'indirect-selection4773' into profiles-global-configs

matthewshaver committed:
Merge branch 'jerco/flags-behavior-changes' into indirect-selection4773

matthewshaver committed:
Merge branch 'indirect-selection4773' into profiles-global-configs

matthewshaver committed:
Merge branch 'current' into jerco/flags-behavior-changes

matthewshaver committed:
Merge branch 'jerco/flags-behavior-changes' into indirect-selection4773

matthewshaver committed:
Merge branch 'indirect-selection4773' into profiles-global-configs

matthewshaver committed:
Profiles global configs (#4992)

## What are you changing in this pull request and why?

Updates the profiles.yml page to reflect the changes to global configs

## Checklist
<!--
Uncomment when publishing docs for a prerelease version of dbt:
- [ ] Add versioning components, as described in [Versioning
Docs](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-entire-pages)
- [ ] Add a note to the prerelease version [Migration
Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)
-->
- [ ] Review the [Content style
guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md)
so my content adheres to these guidelines.
- [ ] For [docs
versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#about-versioning),
review how to [version a whole
page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version)
and [version a block of
content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content).
- [ ] Add a checklist item for anything that needs to happen before this
PR is merged, such as "needs technical review" or "change base branch."

Adding or removing pages (delete if not applicable):
- [ ] Add/remove page in `website/sidebars.js`
- [ ] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/static/_redirects`
- [ ] Run link testing locally with `npm run build` to update the links
that point to deleted pages

matthewshaver committed:
Adding new indirect selection pages (#4977)

resolves #4773 

## What are you changing in this pull request and why?
<!---
Describe your changes and why you're making them. If related to an open 
issue or a pull request on dbt Core, then link to them here! 

To learn more about the writing conventions used in the dbt Labs docs,
see the [Content style
guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md).
-->

## Checklist
<!--
Uncomment when publishing docs for a prerelease version of dbt:
- [ ] Add versioning components, as described in [Versioning
Docs](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-entire-pages)
- [ ] Add a note to the prerelease version [Migration
Guide](https://github.com/dbt-labs/docs.getdbt.com/tree/current/website/docs/docs/dbt-versions/core-upgrade)
-->
- [ ] Review the [Content style
guide](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/content-style-guide.md)
so my content adheres to these guidelines.
- [ ] For [docs
versioning](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#about-versioning),
review how to [version a whole
page](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#adding-a-new-version)
and [version a block of
content](https://github.com/dbt-labs/docs.getdbt.com/blob/current/contributing/single-sourcing-content.md#versioning-blocks-of-content).
- [ ] Add a checklist item for anything that needs to happen before this
PR is merged, such as "needs technical review" or "change base branch."

Adding or removing pages (delete if not applicable):
- [ ] Add/remove page in `website/sidebars.js`
- [ ] Provide a unique filename for new pages
- [ ] Add an entry for deleted pages in `website/static/_redirects`
- [ ] Run link testing locally with `npm run build` to update the links
that point to deleted pages

matthewshaver committed:
Update sidebars.js

Adding page to sidebar

matthewshaver committed:
Update about-global-configs.md

Preventing page loop

matthewshaver committed:
Merge branch 'current' into jerco/flags-behavior-changes

matthewshaver committed:
Update sidebars.js

Fixing spacing

dbeatty10 commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,98 @@
 
-Each config is prioritized over the previous one. For example, if all three are provided, then the CLI flag takes precedence.
+There are multiple ways of setting flags, which depend on the use case:
+- **[Project-level `flags` (`dbt_project.yml`)](/reference/global-configs/project-flags):** Define version-controlled defaults for everyone running this project. Preserve legacy behaviors until their slated deprecation.
```
Nit -- this suggestion would be a little easier on my eyes.

Also linked to the "legacy behaviors" page at the same time.

```suggestion
- **[Project-level `flags` in `dbt_project.yml`](/reference/global-configs/project-flags):** Define version-controlled defaults for everyone running this project. Preserve [legacy behaviors](/reference/global-configs/legacy-behaviors) until their slated deprecation.
```

Here's how it looks currently:
<img width="540" alt="image" src="https://github.com/dbt-labs/docs.getdbt.com/assets/44704949/b603a678-6ed8-40ff-bee8-0fe7d8798054">

dbeatty10 commented on website/docs/reference/global-configs/about-global-configs.md:
```diff
@@ -1,18 +1,98 @@
+
+There are two categories of exceptions:
+1. **Flags setting file paths:** Flags for file paths that are relevant to runtime execution (for example, `--log-path` or `--state`) cannot be set in `dbt_project.yml`. To override defaults, pass CLI options or set environment variables (`DBT_LOG_PATH`, `DBT_STATE`). Flags that tell dbt where to find project resources (for example, `model-paths`) are set in `dbt_project.yml`, but as a top-level  ...
+2. Flags opting into legacy dbt behaviors can _only_ be defined in `dbt_project.yml`. These are intended to be set in version control and migrated via pull/merge request. Their values should not diverge indefinitely across invocations, environments, or users.
```
Nit -- 1) above has some bold text, so it seems like this should too.

Also linked to the "legacy behaviors" page at the same time.

```suggestion
2. **Opt-in flags:** Flags opting into [legacy dbt behaviors](/reference/global-configs/legacy-behaviors) can _only_ be defined in `dbt_project.yml`. These are intended to be set in version control and migrated via pull/merge request. Their values should not diverge indefinitely across invocations, environments, or users.
```

dbeatty10 commented on website/docs/reference/global-configs/project-flags.md:
```diff
@@ -0,0 +1,47 @@
+
+</File>
+
+Reference the [table of all flags](/reference/global-configs/about-global-configs#available-flags) to see which global configs are available for setting in `dbt_project.yml`.
```
Nit -- would be nice to link to docs for `dbt_project.yml`

```suggestion
Reference the [table of all flags](/reference/global-configs/about-global-configs#available-flags) to see which global configs are available for setting in [`dbt_project.yml`](/reference/dbt_project.yml).
```

dbeatty10 commented on website/docs/docs/core/connect-data-platform/profiles.yml.md:
```diff
@@ -54,6 +63,37 @@ This section identifies the parts of your `profiles.yml` that aren't specific to
+</VersionBlock>
+
 ## User config
 
-You can set default values of global configs for all projects that you run using your local machine. See the docs on [global configs](/reference/global-configs/about-global-configs) for details.
+You can set default values of global configs for all projects that you run using your local machine. Refer to [About global configs](/reference/global-configs/about-global-configs) for details.
```
Fast-follow: we should consider adding versioning block(s) here since there will no longer way to set configs/flags across all projects after this is deprecated.

dbeatty10 replied:
Within that fast-follow (or separately), we should consider separating the code example for `config:` within `profiles.yml` entirely. That might make the versioning blocks cleaner and easier to read and maintain.

matthewshaver replied:
Raised in issue https://github.com/dbt-labs/docs.getdbt.com/issues/4996

dbeatty10 commented on website/docs/reference/global-configs/legacy-behaviors.md:
```diff
@@ -0,0 +1,36 @@
+
+## Behaviors
+
+### `source_freshness_run_project_hooks`
+- Should the 'dbt source freshness' command include [on-run-start / on-run-end hooks](/reference/project-configs/on-run-start-on-run-end)?
+- Planned introduction in v1.8, with default 'True'
+- Planned removal in v1.9
```
Fast-follow: this section is hard for me to read and understand.

I'd also expect:
- `dbt source freshness` instead of 'dbt source freshness'
- `True` instead of 'True'

matthewshaver replied:
Raised the issue https://github.com/dbt-labs/docs.getdbt.com/issues/4996

dbeatty10 reviewed (approved):
This has such great info! 🧠

5 non-blocking comments -- any can be converted to fast-follows.
- (3) small suggestions
- (2) fast-follow nudges

matthewshaver committed:
Apply suggestions from code review

Co-authored-by: Doug Beatty <44704949+dbeatty10@users.noreply.github.com>

matthewshaver committed:
Merge branch 'current' into jerco/flags-behavior-changes
